libfwupdplugin/fu-mtd-device.c

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright (C) 2021 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuMtdDevice"
+
+#include "config.h"
+
+#ifdef HAVE_MTD_USER_H
+#include <mtd/mtd-user.h>
+#endif
+
+#include "fu-mtd-device.h"
+
+/**
+ * FuMtdDevice
+ *
+ * A MTD storage device, typically a NAND or SPI chip.
+ *
+ * See also: #FuUdevDevice
+ */
+
+typedef struct {
+	guint64 erasesize;
+	gboolean powerup_lock;
+} FuMtdDevicePrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE(FuMtdDevice, fu_mtd_device, FU_TYPE_UDEV_DEVICE)
+
+#define GET_PRIVATE(o) (fu_mtd_device_get_instance_private(o))
+
+static void
+fu_mtd_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuMtdDevice *self = FU_MTD_DEVICE(device);
+	FuMtdDevicePrivate *priv = GET_PRIVATE(self);
+	fu_common_string_append_kx(str, idt, "EraseSize", priv->erasesize);
+	fu_common_string_append_kb(str, idt, "PowerupLock", priv->powerup_lock);
+}
+
+/**
+ * fu_mtd_device_get_powerup_lock:
+ * @self: a #FuMtdDevice
+ *
+ * Gets the powerup lock state.
+ *
+ * Returns: %TRUE if set
+ *
+ * Since: 1.7.2
+ **/
+gboolean
+fu_mtd_device_get_powerup_lock(FuMtdDevice *self)
+{
+	FuMtdDevicePrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_MTD_DEVICE(self), FALSE);
+	return priv->powerup_lock;
+}
+
+static gboolean
+fu_mtd_device_probe(FuDevice *device, GError **error)
+{
+	FuMtdDevice *self = FU_MTD_DEVICE(device);
+	FuMtdDevicePrivate *priv = GET_PRIVATE(self);
+	const gchar *name;
+	guint64 flags = 0;
+	guint64 size = 0;
+
+	/* FuUdevDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_mtd_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	/* set physical ID */
+	if (!fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "spi,mtd", error))
+		return FALSE;
+
+	/* get name */
+	name = fu_udev_device_get_sysfs_attr(FU_UDEV_DEVICE(device), "name", NULL);
+	if (name != NULL) {
+		g_autofree gchar *devid = NULL;
+		g_autofree gchar *name_safe = g_strdup(name);
+		g_strdelimit(name_safe, " /\\\"", '-');
+		devid = g_strdup_printf("MTD\\NAME_%s", name_safe);
+		fu_device_add_instance_id(FU_DEVICE(self), devid);
+	}
+
+	/* get properties about the device */
+	if (!fu_udev_device_get_sysfs_attr_uint64(FU_UDEV_DEVICE(device), "size", &size, NULL))
+		return FALSE;
+	fu_device_set_firmware_size_max(device, size);
+	if (!fu_udev_device_get_sysfs_attr_uint64(FU_UDEV_DEVICE(device), "flags", &flags, NULL))
+		return FALSE;
+#ifdef HAVE_MTD_USER_H
+	if ((flags & MTD_NO_ERASE) == 0) {
+		if (!fu_udev_device_get_sysfs_attr_uint64(FU_UDEV_DEVICE(device),
+							  "erasesize",
+							  &priv->erasesize,
+							  NULL))
+			return FALSE;
+	}
+	if (flags & MTD_WRITEABLE)
+		fu_device_add_flag(device, FWUPD_DEVICE_FLAG_UPDATABLE);
+	if (flags & MTD_POWERUP_LOCK)
+		priv->powerup_lock = TRUE;
+#endif
```

Comment:
I think you should fail probe without that `HAVE_MTD_USER_H` no?  It seems pointless to me to show the device as around but not updatable.

Reply:
I think it's best to disable the `mtd` plugin (not yet pushed) in this case; this means we don't do the udev probing stuff at all and it seems cleaner.